internal/webhook/external/handlers/refresh_warehouse.go

```diff
@@ -0,0 +1,119 @@
+package handlers
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/api"
+	"github.com/akuity/kargo/internal/indexer"
+	"github.com/akuity/kargo/internal/webhook/external/providers"
+)
+
+// NewRefreshWarehouseWebhook handles push events for the designated provider.
+// After the provider has been resolved and the request has been authenticated,
+// the kubeclient is queried for all warehouses that contain a subscription
+// to the repo in question. Those warehouses are then patched with a special
+// annotation that signals down stream logic to refresh the warehouse.
+func (f *Factory) NewRefreshWarehouseWebhook(name providers.Name) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		f.log.Info("initializing provider")
+		p, err := providers.New(name)
+		if err != nil {
+			// If there's a missing secret/env var
+			// we don't want to leak that information
+			// from our API so just log the part that's
+			// useful to us and return a generic error
+			// to the client.
+			f.log.Error(err, "failed to initialize provider")
+			http.Error(w,
+				"failed to initialize provider",
+				http.StatusInternalServerError,
+			)
+			return
+		}
+
+		f.log.Info("provider initialized",
+			"provider", name.String(),
+		)
```

Comment:
It would be quite a bit more straightforward to inject the provider directly.

```suggestion
func (f *Factory) NewRefreshWarehouseWebhook(p Provider) http.HandlerFunc {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```